docs/contributing/git-commit-guide.md

```diff
@@ -0,0 +1,90 @@
+
+Each commit should represent a single logical change. A commit should:
+
+- Contain only related modifications.
+- Be small enough to understand but large enough to encompass a meaningful unit of change.
+- Avoid including unrelated changes, such as formatting updates along with functional modifications.
+
+### 2. Split Large Changes
+
+If a change touches multiple aspects of the code or is too large to be reviewed easily, break it into multiple smaller
+commits:
+
+- Separate refactoring from feature implementation.
+- Isolate bug fixes from new features.
+- Ensure that each commit is functional and does not leave the repository in a broken state.
+
+#### Rule of Thumb for Large Commits
+
+- If a commit changes more than **200–300 lines of code**, consider splitting it.
+- If multiple files across **different components or concerns** are modified, break them into separate commits.
+- If reviewing the commit would take **longer than 10–15 minutes**, it's likely too big.
+
+### 3. Use Consistent Commit Message Naming
+
+Commit messages should start with a clear action verb that describes the change:
+
+- **`Add`**: For new features, tests, or documentation.
+- **`Change`**: For modifications to existing functionality.
+- **`Fix`**: For bug fixes and patches.
+- **`Remove`**: For deletions or removals of code or dependencies.
+- **`Refactor`**: For restructuring code without changing functionality.
+- **`Bump`**: For dependency version updates.
+
+**Example:**
+
+```
+Add validation to login input to prevent SQL injection
+
+Change login input validation to sanitize user inputs properly before being processed. Prevents potential SQL injection vulnerabilities.
+
+Fixes #123.
+```
+
+### 4. Write Descriptive Commit Messages
+
+A good commit message follows this structure:
+
+- **Title (short summary, 50 characters max)**
+- **Body (optional, but recommended for context)**
+  - Explain *what* the change does and *why* it was needed.
+  - Describe any important details, such as edge cases addressed.
+  - Reference relevant issue numbers or discussions.
+
+Avoid generic descriptions like "Fix bug" or "Update code"
+
+### 5. Ensure Commits Are Self-Contained
+
+Each commit should:
+
+- Be complete and independently buildable.
+- Not introduce temporary hacks or broken functionality.
+- Ideally, pass tests to maintain project integrity.
+
+### 6. Help Reviewers and Future Developers
+
+Consider the commit history as a documentation tool. Someone reviewing past changes should be able to:
+
+- Follow the logical evolution of a feature or fix.
+- Understand *how* and *why* a functionality changed over time.
+- Easily revert a commit if needed without breaking unrelated functionality.
+
+### 7. Use Atomic Commits
+
+Atomic commits mean that each commit represents a single unit of change that can be reverted independently. This
+practice ensures that if an issue arises, it is easier to pinpoint and undo the specific problematic change.
```

Comment:
Based on how we've worked before I had the understanding that we're working atomic at the level of a merge, but that individual commits within the merge might not be completely functional.

I think this is worth clarifying. If we really require atomic commits to each individual level, then we might as well do rebases instead of merges because each commit should work as it is and we can just back out the offending commit.

If we're working at the merges level, then while we shouldn't intentionally break between commits, we'd be backing out the merge and bisecting based on merges.

There is also some overlap between (6) and (2) where we're again saying it needs to work inbetween commits.

Reply:
This doesn't mean to have something completely functional, but it should compile, tests should succeed and all quality tools issues should be resolved. This is meant to prevent commits like:

1. Add funtionality
2. Add test for functionality
3. Fix broken tests
4. Fix linter issues
5. Fix code formatting
6. ...

These changes should be part of one commit.

Reply:
This is a special case. If it is detectable using our build, quality checks and tests then yes, otherwise I think it's fine.